listen_ports_facts: Added support for ss

moonrail opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
This PR adds support for command `ss` to module `listen_ports_facts`.

Some distros do not include `netstat` in their base installs anymore, whereas `ss` is included.

Default behaviour is extended to try supported commands one by one until one is existing on the machine.
`netstat` is tried first, as it is the current default and `ss` returns all processes per listen address & port.

A user can override the auto-detection by passing the argument `command` for either `netstat` or `ss`.

Modified integration tests accordingly and tested with them successfully on:
* Debian 9, 10, 11
* Ubuntu 16.04, 18.04, 20.04
* CentOS 7, 8, Stream
* RedHat 7, 8

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
listen_ports_facts

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below
<no output changes>
```

russoz commented:
Hi @moonrail ,
Thank you for the contribution! Please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to).

Other than the nits picked by the sanity tests, it looks good to me.

felixfontein commented on changelogs/fragments/3708-listen_ports_facts-add-ss-support.yml:
```diff
@@ -0,0 +1,3 @@
+---
+minor_changes:
+  - listen_ports_facts - add support for `ss` command besides `netstat` (https://github.com/ansible-collections/community.general/pull/3708)
```
```suggestion
  - listen_ports_facts - add support for ``ss`` command besides ``netstat`` (https://github.com/ansible-collections/community.general/pull/3708).
```

felixfontein commented on plugins/modules/system/listen_ports_facts.py:
```diff
@@ -13,11 +13,25 @@
+    required: false
+    choices:
+      - netstat
+      - ss
```
```suggestion
      - ss
    version_added: 4.1.0
```

felixfontein commented on plugins/modules/system/listen_ports_facts.py:
```diff
@@ -13,11 +13,25 @@
+      - Override which command to use for fetching listen ports.
+      - 'By default module will use first found supported command on the system (in order of documentation)'
+    type: str
+    required: false
```
```suggestion
```

felixfontein commented on plugins/modules/system/listen_ports_facts.py:
```diff
@@ -13,11 +13,25 @@
+  command:
+    description:
+      - Override which command to use for fetching listen ports.
+      - 'By default module will use first found supported command on the system (in order of documentation)'
```
```suggestion
      - 'By default module will use first found supported command on the system (in order of documentation).'
```

felixfontein commented on plugins/modules/system/listen_ports_facts.py:
```diff
@@ -13,11 +13,25 @@
 short_description: Gather facts on processes listening on TCP and UDP ports.
+notes:
+  - |
+    ss returns all processes for each listen address & port.
+    This plugin will return each of them, so multiple entries for the same listen address & port are likely in results
```
```suggestion
    C(ss) returns all processes for each listen address and port.
    This plugin will return each of them, so multiple entries for the same listen address and port are likely in results.
```

felixfontein commented on plugins/modules/system/listen_ports_facts.py:
```diff
@@ -13,11 +13,25 @@
     - Nathan Davison (@ndavison)
 description:
-    - Gather facts on processes listening on TCP and UDP ports using netstat command.
+    - Gather facts on processes listening on TCP and UDP ports using netstat or ss command.
```
```suggestion
    - Gather facts on processes listening on TCP and UDP ports using the C(netstat) or C(ss) commands.
```

felixfontein commented on plugins/modules/system/listen_ports_facts.py:
```diff
@@ -181,10 +195,75 @@ def netStatParse(raw):
     module = AnsibleModule(
-        argument_spec={},
+        argument_spec=dict(
+            command=dict(type='str', required=False, choices=list(commands_map.keys()))
```
```suggestion
            command=dict(type='str', choices=list(commands_map))
```
(The `.keys()` isn't needed, by default dictionaries iterate over their keys.)

moonrail replied:
@felixfontein 
Changed as you've requested.
I just like explicit usage of "keys()" as it gives you at first glance hints on the data type of a variable and what is iterated.

felixfontein commented on plugins/modules/system/listen_ports_facts.py:
```diff
@@ -220,18 +299,34 @@ def getPidUser(pid):
+            command = module.params['command']
+            bin_path = module.get_bin_path(command, required=True)
+        else:
+            for c in commands_map.keys():
```
```suggestion
            for c in sorted(commands_map):
```
Dictionary order is not fixed, so you cannot rely on it. (It is only fixed for Python 3.5 or 3.6 and later; I forgot which of the two.)

moonrail replied:
@felixfontein 
Good catch, thank you!

felixfontein commented on plugins/modules/system/listen_ports_facts.py:
```diff
@@ -220,18 +299,34 @@ def getPidUser(pid):
+                    break
+
+        if bin_path is None:
+            raise EnvironmentError(msg='Unable to find any of the supported commands in PATH: {0}'.format(", ".join(commands_map.keys())))
```
```suggestion
            raise EnvironmentError(msg='Unable to find any of the supported commands in PATH: {0}'.format(", ".join(sorted(commands_map))))
```

felixfontein commented on plugins/modules/system/listen_ports_facts.py:
```diff
@@ -220,18 +299,34 @@ def getPidUser(pid):
         # which ports are listening for connections?
-        rc, stdout, stderr = module.run_command([netstat_cmd, '-plunt'])
+        args = commands_map[command]['args']
+        rc, stdout, stderr = module.run_command([bin_path, args])
```
Since you are using `args` and not `arg`, you want to change `commands_map` to provide a list and not a string, and replace this with
```suggestion
        rc, stdout, stderr = module.run_command([bin_path] + args)
```

moonrail commented:
Hi @russoz,
I've added the fragment - totally forgot it.
Your automatic tests showed, that `ss` will not always output the column "process", even with root permissions, therefore I've made changes to support it.
For other possible surprises I've added rudimentary validation and descriptive error messages.

Tests look good now.

russoz commented:
LGTM after these last adjustments from Felix.

felixfontein commented on plugins/modules/system/listen_ports_facts.py:
```diff
@@ -181,10 +195,87 @@ def netStatParse(raw):
+    lines = lines[1:]
+
+    for line in lines:
+        cells = line.split()
```
Is it possible that `process` contains spaces? In that case, having a value for `maxsplit` would be a good idea:
```suggestion
        cells = line.split(None, 6)
```

russoz replied:
I would be very surprised if that's a real thing, but not sure if OS'es forbid it in any way.

moonrail replied:
I've never seen a process name with spaces or any whitespace char on any linux distro, so I would also be very surprised.
But having maxsplit does not harm anything, so I'll integrate it.

felixfontein commented on plugins/modules/system/listen_ports_facts.py:
```diff
@@ -13,11 +13,25 @@
+notes:
+  - |
+    C(ss) returns all processes for each listen address & port.
+    This plugin will return each of them, so multiple entries for the same listen address and port are likely in results
```
```suggestion
    This plugin will return each of them, so multiple entries for the same listen address and port are likely in results.
```

felixfontein commented on plugins/modules/system/listen_ports_facts.py:
```diff
@@ -13,11 +13,25 @@
 short_description: Gather facts on processes listening on TCP and UDP ports.
+notes:
+  - |
+    C(ss) returns all processes for each listen address & port.
```
```suggestion
    C(ss) returns all processes for each listen address and port.
```

felixfontein commented:
@moonrail thanks for implementing this!
@russoz thanks for reviewing!
